internal/eventstream/client_api.go

```diff
@@ -0,0 +1,421 @@
+func (client *Client) connectEventStream(esTypes []string) (*http.Response, context.CancelFunc, error) {
+	apiUrl, err := url.JoinPath(client.ApiHost, "/v1/events")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for i := 0; ; i++ {
+		// Always ensure an unique queue name to ensure no conflicts might occur.
+		queueNameRndBuff := make([]byte, 16)
+		_, _ = rand.Read(queueNameRndBuff)
+
+		reqBody, err := json.Marshal(map[string]any{
+			"queue": fmt.Sprintf("icinga-notifications-%x", queueNameRndBuff),
+			"types": esTypes,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Sub-context which might get canceled early if connecting takes to long.
+		// The reqCancel function will be called after the select below or when leaving the function with an error.
+		reqCtx, reqCancel := context.WithCancel(client.Ctx)
+
+		req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, apiUrl, bytes.NewReader(reqBody))
+		if err != nil {
+			reqCancel()
+			return nil, nil, err
+		}
+
+		req.SetBasicAuth(client.ApiBasicAuthUser, client.ApiBasicAuthPass)
+		req.Header.Set("Accept", "application/json")
+		req.Header.Set("Content-Type", "application/json")
+
+		resCh := make(chan *http.Response)
+
+		go func() {
+			defer close(resCh)
+
+			client.Logger.Debug("Try to establish an Event Stream API connection")
+			httpClient := &http.Client{Transport: &client.ApiHttpTransport}
+			res, err := httpClient.Do(req)
+			if err != nil {
+				client.Logger.Warnw("Establishing an Event Stream API connection failed, will be retried", zap.Error(err))
+				return
+			}
+
+			select {
+			case <-reqCtx.Done():
+				// This case might happen when this httpClient.Do and the time.After in the select below finish at round
+				// about the exact same time, but httpClient.Do was slightly faster than reqCancel().
+				_, _ = io.Copy(io.Discard, res.Body)
+				_ = res.Body.Close()
+			case resCh <- res:
+			}
+		}()
+
+		select {
+		case res, ok := <-resCh:
+			if ok {
+				return res, reqCancel, nil
+			}
+
+		case <-time.After(3 * time.Second):
+		}
+		reqCancel()
+
+		// Rate limit API reconnections: slow down for successive failed attempts but limit to three seconds.
+		time.Sleep(slices.Min([]time.Duration{3 * time.Second, time.Duration(i*200) * time.Millisecond}))
+	}
+}
+
+// listenEventStream subscribes to the Icinga 2 API Event Stream and handles received objects.
+//
+// In case of a parsing or handling error, this error will be returned. If the server closes the connection, nil will
+// be returned.
+func (client *Client) listenEventStream() error {
+	// Ensure to implement a handler case in the type switch below for each requested type.
+	response, cancel, err := client.connectEventStream([]string{
+		typeStateChange,
+		typeAcknowledgementSet,
+		// typeAcknowledgementCleared,
+		// typeCommentAdded,
+		// typeCommentRemoved,
+		// typeDowntimeAdded,
+		// typeDowntimeRemoved,
+		// typeDowntimeStarted,
+		// typeDowntimeTriggered,
+	})
```

Comment:
Yes, in theory I would agree. However, in this very case creating a local sub-context is kind of a problem, as this context is only used to allow aborting the web request after a timeout. Thus, the local `context.CancelFunc` is only called in case of an early error, as the Event Stream connection is bound to it.

When leaving the `connectEventStream` method with an established connection, the context is still in use. But in this case the `context.CancelFunc` was not called and would go out of scope.

Against this, the `context.CancelFunc` needs still to be returned - as in the current implementation -, resulting in an even more oddly looking pattern with then two `context.CancelFunc`s to be called. Like so:
```go
	ctx, cancel := context.WithCancel(client.Ctx)
	defer cancel()

	response, rseponseCancel, err := client.connectEventStream(ctx, []string{ … })
	if err != nil { … }
	defer func() {
		cancel()

		_, _ = io.Copy(io.Discard, response.Body)
		_ = response.Body.Close()
	}()
```

As an alternative, one could rely on our "domain knowledge" and just accept that the method-local `context.CancelFunc` will not be called directly, but as being derived from a context which will eventually be canceled. This, however, will look even more unconventional, might result in future bugs when this knowledge - documented or not - gets forgotten, and will trigger each linter. In fact, as I have just played with this idea, even GoLand's quite conservative linter lighted up this case.

To my knowledge, there are only two reasonable ways to terminate a web request early: by timeouts, which is a bit tricky for long-lived connections and which was already discussed <https://github.com/Icinga/icinga-notifications/pull/112#discussion_r1376514352>, and by a `context.Context`.

That's why I would stay with the current architecture for the moment. However, I'm open to other suggestions.

Reply:
> This, however, will look even more unconventional, might result in future bugs when this knowledge - documented or not - gets forgotten, and will trigger each linter. In fact, as I have just played with this idea, even GoLand's quite conservative linter lighted up this case.

Okay, I didn't consider that part. (Would be kind of nice if there was some way to cancel the `CancelFunc`, so that the child context becomes pretty much the same as its parent. But I don't know of any such way except implementing a new `context.Context`, so probably not.)

> However, I'm open to other suggestions.

I think my suggestion was mainly caused by the defer function having to deal with HTTP and context internals of `connectEventStream()` about which we don't really care here. So another quick idea: maybe just return an `io.ReadCloser` that just does all the right cleanup on both read errors and close. (Needs implementing an extra type though, so not fully sure how much nicer this will be in the end.)

Reply:
I liked the `io.ReadCloser` idea and went with it in c305b9e0ce548109396fe681bfd3506dc2ea47ef.